File: backend/app/Models/Principal.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;

/**
 * Class Principal.
 *
 * @OA\Schema(
 *     description="Principal model",
 *     title="Principal model",
 *     required={"tconst", "nconst", "category"},
 *     @OA\Xml(
 *         name="Principal"
 *     )
 * )
 */
class Principal extends Model
{
    use HasFactory;

    public $timestamps = false;

    /**
     * @var array
     */
    protected $fillable = ['tconst', 'nconst', 'category', 'job', 'characters', 'img_url_asset'];

    protected $casts = [
        'characters' => 'array',
    ];

    public function person()
    {
        return $this->belongsTo(Person::class, 'nconst');
    }

    public function movie()
    {
        return $this->belongsTo(Movie::class, 'tconst');
    }
    /**
     * @OA\Property(
     *     format="string",
     *     description="tconst",
     *     title="tconst",
     * )
     *
     * @var string
     */
    private $tconst;
    /**
     * @OA\Property(
     *     format="string",
     *     description="nconst",
     *     title="nconst",
     * )
     *
     * @var string
     */
    private $nconst;
    /**
     * @OA\Property(
     *     format="string",
     *     description="category",
     *     title="category",
     * )
     *
     * @var string
     */
    private $category;
    /**
     * @OA\Property(
     *     format="string",
     *     description="job",
     *     title="job",
     * )
     *
     * @var string
     */
    private $job;
    /**
     * @OA\Property(
     *     format="string",
     *     description="characters",
     *     title="characters",
     * )
     *
     * @var string
     */
    private $characters;
    /**
     * @OA\Property(
     *     format="string",
     *     description="img_url_asset",
     *     title="img_url_asset",
     * )
     *
     * @var string
     */
    private $img_url_asset;
}

File: frontend/lib/shared/models/principal.dart
import 'dart:convert';

import 'package:ntuaflix/shared/models/person.dart';

class Principal {
  final int id;
  final String tconst;
  final int ordering;
  final String nconst;
  final String category;
  final String? job;
  final String? characters;
  final String? imgUrlAsset;
  final Person? person;

  Principal({
    required this.id,
    required this.tconst,
    required this.ordering,
    required this.nconst,
    required this.category,
    required this.job,
    required this.characters,
    required this.imgUrlAsset,
    required this.person,
  });

  Principal copyWith({
    int? id,
    String? tconst,
    int? ordering,
    String? nconst,
    String? category,
    String? job,
    String? characters,
    String? imgUrlAsset,
    Person? person,
  }) =>
      Principal(
        id: id ?? this.id,
        tconst: tconst ?? this.tconst,
        ordering: ordering ?? this.ordering,
        nconst: nconst ?? this.nconst,
        category: category ?? this.category,
        job: job ?? this.job,
        characters: characters ?? this.characters,
        imgUrlAsset: imgUrlAsset ?? this.imgUrlAsset,
        person: person ?? this.person,
      );

  factory Principal.fromJson(String str) => Principal.fromMap(json.decode(str));

  String toJson() => json.encode(toMap());

  factory Principal.fromMap(Map<String, dynamic> json) => Principal(
        id: json["id"],
        tconst: json["tconst"],
        ordering: json["ordering"],
        nconst: json["nconst"],
        category: json["category"],
        job: json["job"],
        characters: json["characters"],
        imgUrlAsset: json["img_url_asset"],
        person: json["person"] != null ? Person.fromMap(json["person"]) : null,
      );

  Map<String, dynamic> toMap() => {
        "id": id,
        "tconst": tconst,
        "ordering": ordering,
        "nconst": nconst,
        "category": category,
        "job": job,
        "characters": characters,
        "img_url_asset": imgUrlAsset,
        "person": person?.toMap(),
      };
}
